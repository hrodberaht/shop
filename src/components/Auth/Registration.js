import React, { Component } from 'react';
import { Field, reduxForm, SubmissionError } from 'redux-form';
import { Link, withRouter } from 'react-router-dom';
import * as yup from 'yup';
import dataFetcher from '../../shared/dataFetcher';

const schema = yup.object().shape({
  email: yup
    .string()
    .email()
    .required(),
  firstName: yup.string().required(),
  lastName: yup.string().required(),
  password: yup
    .string()
    .required()
    .min(6),
  companyId: yup.string().required(),
});

const asyncValidate = values => schema.validate(values, { abortEarly: false }).catch((err) => {
  const errorsForm = {};
  err.inner.forEach((error) => {
    errorsForm[error.path] = error.type;
  });
  throw errorsForm;
});

export class Registration extends Component {
  state = {
    companies: [],
    loaded: false,
    error: false,
  };

  componentDidMount() {
    this.fetchCompanies();
  }

  fetchCompanies = () => dataFetcher('companies', 'get')
    .then((data) => {
      this.setState({ companies: data, loaded: true });
    })
    .catch(() => {
      this.setState({ error: 'Server not working' });
    });

  addUserToDB = (values) => {
    const errorsForm = {};
    return dataFetcher('registration', 'post', null, values).then((res) => {
      if (res.status === 422) {
        errorsForm.email = res.error;
        return errorsForm;
      }

      return res;
    });
  };

  submit = async (values) => {
    const check = await this.addUserToDB(values);
    if (check.error) {
      return Promise.resolve().then(() => {
        throw new SubmissionError({ email: check.error });
      });
    }
    alert('User added');
    return this.props.history.push('/');
  };

  renderField = ({
    input, label, type, meta: { touched, error },
  }) => (
    <div>
      <input className="form-input" {...input} type={type} placeholder={label} />
      <p className="error-text">{touched && (error && <span>{error}</span>)}</p>
    </div>
  );

  selectOptions = () => (
    <React.Fragment>
      <option />
      {this.state.companies.map(company => (
        <option key={company.id} value={company.id}>
          {company.name}
        </option>
      ))}
    </React.Fragment>
  );

  render() {
    const { handleSubmit, pristine } = this.props;
    if (this.state.error) return <div>{this.state.error}</div>;
    return (
      <div className="registration-form">
        <form onSubmit={handleSubmit(this.submit)}>
          <label className="form-label" htmlFor="email">
            <p>Email:</p>
            <Field name="email" type="text" component={this.renderField} label="Email" />
          </label>
          <label className="form-label" htmlFor="first-name">
            <p>First name:</p>
            <Field name="firstName" type="text" component={this.renderField} label="First name" />
          </label>
          <label className="form-label" htmlFor="last-name">
            <p>Last name:</p>
            <Field name="lastName" type="text" component={this.renderField} label="Last name" />
          </label>
          <label className="form-label" htmlFor="password">
            <p>Password:</p>
            <Field name="password" type="password" component={this.renderField} label="Password" />
          </label>
          <label className="form-label" htmlFor="password">
            <p>Your Company</p>
            <Field className="registration-select" name="companyId" component="select">
              {this.state.loaded && this.selectOptions()}
            </Field>
          </label>
          <button className="btn btn-submit" type="submit" disabled={pristine}>
            Submit
          </button>
          <div className="registration-link">
            <Link to="/">Login</Link>
          </div>
        </form>
      </div>
    );
  }
}

withRouter(Registration);
export default reduxForm({
  asyncValidate,
  form: 'registration',
})(Registration);
