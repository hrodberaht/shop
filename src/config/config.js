const config = {
  url: 'http://localhost:3004/',
  graphql: 'http://localhost:4000/graphql',
};

export default config;
