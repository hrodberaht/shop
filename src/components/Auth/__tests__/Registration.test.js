import fetchMock from 'fetch-mock/es5/client';
import { Registration } from '../Registration';
import config from '../../../config/config';

const defaultProps = {
  handleSubmit: jest.fn(),
};
const setup = buildSetup(Registration, defaultProps);

afterEach(() => {
  fetchMock.restore();
});

describe('<Registration />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('fetchCompanies set error in state if server not working', async () => {
    const { wrapper } = setup();
    const instance = wrapper.instance();
    fetchMock.get(`${config.url}companies`, { throws: Error });
    await instance.fetchCompanies();
    expect(wrapper.state().error).toBe('Server not working');
  });

  it('fetchCompanies set companies in state, and loaded to true ', async () => {
    const { wrapper } = setup();
    const instance = wrapper.instance();
    const companies = [
      {
        id: 'company-1',
        name: 'IBM',
      },
    ];
    fetchMock.get(`${config.url}companies`, {
      headers: { 'content-type': 'application/json' },
      body: companies,
    });
    await instance.fetchCompanies();
    expect(wrapper.state().companies).toEqual(companies);
    expect(wrapper.state().loaded).toEqual(true);
  });

  it('addUserToDB should return error if status 422 ', async () => {
    const { wrapper } = setup();
    const instance = wrapper.instance();
    const error = { error: 'email is taken' };
    fetchMock.post(`${config.url}registration`, {
      status: 422,
      body: error,
    });
    expect(await instance.addUserToDB()).toEqual(error);
  });

  it('addUserToDB should return response', async () => {
    const { wrapper } = setup();
    const instance = wrapper.instance();
    const response = { message: 'user added' };
    fetchMock.post(`${config.url}registration`, {
      status: 422,
      body: response,
    });
    await instance.addUserToDB();
    expect(await instance.addUserToDB()).toEqual(response);
  });

  it('submit should call addUserToDB with values', () => {
    const { wrapper } = setup();
    const instance = wrapper.instance();
    const values = {
      email: 'test@test.com',
      firstName: 'test',
      lastName: 'test',
      password: '123456',
      companyId: 'test',
    };
    jest.spyOn(instance, 'addUserToDB');
    instance.submit(values);
    expect(instance.addUserToDB).toHaveBeenCalledWith(values);
  });

  it('selectOptions() creact select options with companies', () => {
    const { wrapper } = setup();
    const companies = [
      {
        id: 'company-1',
        name: 'IBM',
      },
      {
        id: 'company-2',
        name: 'Apple',
      },
    ];
    const instance = wrapper.instance();
    wrapper.setState({ companies });
    expect(instance.selectOptions()).toMatchSnapshot();
  });

  it('after submit form, call handleSubmit with submit', () => {
    const { wrapper } = setup();
    const instance = wrapper.instance();
    wrapper.find('form').simulate('submit');
    expect(instance.props.handleSubmit).toHaveBeenCalledWith(instance.submit);
  });
});
