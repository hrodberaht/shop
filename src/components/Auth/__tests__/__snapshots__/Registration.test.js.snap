// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> selectOptions() creact select options with companies 1`] = `
<React.Fragment>
  <option />
  <option
    value="company-1"
  >
    IBM
  </option>
  <option
    value="company-2"
  >
    Apple
  </option>
</React.Fragment>
`;

exports[`<Registration /> should render without crash 1`] = `
<div
  className="registration-form"
>
  <form>
    <label
      className="form-label"
      htmlFor="email"
    >
      <p>
        Email:
      </p>
      <Field
        component={[Function]}
        label="Email"
        name="email"
        type="text"
      />
    </label>
    <label
      className="form-label"
      htmlFor="first-name"
    >
      <p>
        First name:
      </p>
      <Field
        component={[Function]}
        label="First name"
        name="firstName"
        type="text"
      />
    </label>
    <label
      className="form-label"
      htmlFor="last-name"
    >
      <p>
        Last name:
      </p>
      <Field
        component={[Function]}
        label="Last name"
        name="lastName"
        type="text"
      />
    </label>
    <label
      className="form-label"
      htmlFor="password"
    >
      <p>
        Password:
      </p>
      <Field
        component={[Function]}
        label="Password"
        name="password"
        type="password"
      />
    </label>
    <label
      className="form-label"
      htmlFor="password"
    >
      <p>
        Your Company
      </p>
      <Field
        className="registration-select"
        component="select"
        name="companyId"
      />
    </label>
    <button
      className="btn btn-submit"
      type="submit"
    >
      Submit
    </button>
    <div
      className="registration-link"
    >
      <Link
        replace={false}
        to="/"
      >
        Login
      </Link>
    </div>
  </form>
</div>
`;
